src/orange/SenseUtils.hh

```diff
@@ -0,0 +1,170 @@
+        : quadric < 0 ? SignedSense::inside
+        : SignedSense::outside;
+ * \endcode
+ * as
+ * \code
+    int gz = !(quadric <= 0) ? 1 : 0;
+    int lz = quadric < 0 ? 1 : 0;
+    return static_cast<SignedSense>(gz - lz);
+ * \endcode
+ * and compressed into a single line.
+ *
+ * NaN values are treated as "outside".
+ */
+[[nodiscard]] CELER_CONSTEXPR_FUNCTION SignedSense
+real_to_sense(real_type quadric)
+{
+    return static_cast<SignedSense>(!(quadric <= 0) - (quadric < 0));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Convert a signed sense to a Sense enum.
+ */
+CELER_CONSTEXPR_FUNCTION Sense to_sense(SignedSense s)
+{
+    return Sense(static_cast<int>(s) >= 0);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Convert a signed sense to a surface state.
+ */
+CELER_CONSTEXPR_FUNCTION SurfaceState to_surface_state(SignedSense s)
+{
+    return s == SignedSense::on ? SurfaceState::on : SurfaceState::off;
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS (HOST)
+//---------------------------------------------------------------------------//
+//! Get a printable character corresponding to a sense.
+inline constexpr char to_char(Sense s)
+{
+    return s == Sense::inside ? '-' : '+';
+}
+
+// Get a string corresponding to a signed sense
+inline char const* to_cstring(SignedSense s)
+{
+    switch (s)
+    {
+        case SignedSense::inside:
+            return "inside";
+        case SignedSense::on:
+            return "on";
+        case SignedSense::outside:
+            return "outside";
+    }
+    return "<invalid>";
+}
+
+//---------------------------------------------------------------------------//
+// CLASSES
+//---------------------------------------------------------------------------//
+/*!
+ * Wrapper for a sense value with associated metadata.
+ */
+class SenseMeta
+{
+  private:
+    enum : size_type
+    {
+        sense_bit,
+        cached_bit,
+    };
+
+  public:
+    constexpr SenseMeta() = default;
+
+    //! Construct with a sense value
+    CELER_CONSTEXPR_FUNCTION SenseMeta(Sense sense)
+    {
+        sense_[sense_bit] = static_cast<bool>(sense);
+        sense_[cached_bit] = true;
+    }
+
+    //! Convert to a sense value
+    CELER_CONSTEXPR_FUNCTION operator Sense() const
+    {
+        return to_sense(sense_[sense_bit]);
+    }
+
+    //! Convert to a boolean value
+    CELER_CONSTEXPR_FUNCTION explicit operator bool() const
+    {
+        return sense_[sense_bit];
+    }
+
+    //! Assign a sense value
+    CELER_CONSTEXPR_FUNCTION SenseMeta& operator=(Sense sense)
+    {
+        sense_[sense_bit] = static_cast<bool>(sense);
+        sense_[cached_bit] = true;
+        return *this;
+    }
+
+    //! Check wether there is a cached sense value
+    CELER_CONSTEXPR_FUNCTION bool cached() const { return sense_[cached_bit]; }
```

Comment:
```suggestion
    CELER_CONSTEXPR_FUNCTION bool is_valid() const { return sense_[cached_bit]; }
```
or `is_assigned`.  The 'thing' might be used as a cache but know nothing on how to calculate/acquire the needed to fill the 'cache'.